packages/flutter/lib/src/widgets/app.dart

```diff
@@ -149,6 +166,54 @@ class WidgetsApp extends StatefulWidget {
   /// for this application's [Localizations] widget.
   final Iterable<LocalizationsDelegate<dynamic>> localizationsDelegates;
 
+  /// This callback is responsible for choosing the app's locale
+  /// when the app is started, and when the user changes the
+  /// device's locale.
+  ///
+  /// The returned value becomes the locale of this app's [Localizations]
+  /// widget. The callback's `locale` parameter is the device's locale when
+  /// the app started, or the device locale the user selected after the app was
+  /// started. The callback's `supportedLocales` parameter is just the value
+  /// [supportedLocales].
+  ///
+  /// If the callback is null or if it returns null then the resolved locale is:
+  /// - The callback's `locale` parameter if it's equal to a supported locale.
+  /// - The first supported locale with the same [Locale.langaugeCode] as the
+  ///   callback's `locale` parameter.
+  /// - The first locale in [supportedLocales].
+  ///
+  /// See also:
+  ///
+  ///  * [MaterialApp.resolveLocaleCallback], which sets the callback of the
+  ///    [WidgetsApp] it creates.
+  final ResolveLocaleCallback resolveLocaleCallback;
+
+  /// The list of locales that this app has been localized for.
+  ///
+  /// By default only the American English locale is supported. Apps should
+  /// configure this list to match the locales they support.
+  ///
+  /// This list must not null. Its default value is just
+  /// `[const Locale('en', 'US')]`.
+  ///
+  /// The order of the list matters. By default, if the device's locale doesn't
+  /// exactly match a supported locale then the first supported locale with a
+  /// matching [Locale.languageCode] is used. If that fails then the first
+  /// supported locale is used. The default locale resolution algorithm can
+  /// be overridden with [resolveLocaleCallback].
+  ///
+  /// See also:
+  ///
+  ///  * [MaterialApp.supportedLocales], which sets the `supportedLocales`
+  ///    of the [WidgetsApp] it creates.
+  ///
+  ///  * [resolveLocaleCallback], an app callback that resolves the app's locale
+  ///    when the device's locale changes.
+  ///
+  ///  * [localizationDelegates], which collectively define all of the localized
+  ///    resources used by this app.
```

Comment:
maybe this should also point to the list of locales that, if listed here, will automatically get a supported translation from us